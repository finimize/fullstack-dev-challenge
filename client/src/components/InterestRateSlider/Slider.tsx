import React, { FC, useContext } from 'react'
import {
    Box,
    Text,
    Slider,
    SliderTrack,
    SliderFilledTrack,
    SliderThumb,
    FormLabel,
} from '@chakra-ui/react'
import { AppContext } from '../../store'
import { UPDATE_INTEREST } from '../../store/types'

export const InterestRateSlider: FC = () => {
    const { state, dispatch } = useContext(AppContext)
    const { calculations } = state

    return (
        <Box padding='2' maxWidth='72' marginBottom='2' marginRight='4'>
            <FormLabel fontSize='sm' marginBottom='2' color='grey5' fontWeight='500'>
                Interest Rate
            </FormLabel>
            <Box display='flex' alignItems='center'>
                <Text fontSize='xs' color='grey5'>
                    0%
                </Text>
                <Slider
                    defaultValue={0}
                    value={calculations.interestRate}
                    min={0}
                    max={50}
                    step={5}
                    margin={4}
                    focusThumbOnChange={false}
                    onChange={(e) =>
                        dispatch({
                            type: UPDATE_INTEREST,
                            payload: Number(e),
                        })
                    }
                >
                    <SliderTrack bg='red.100'>
                        <Box position='relative' right={10} />
                        <SliderFilledTrack bg='tomato' />
                    </SliderTrack>
                    <SliderThumb boxSize={6} />
                </Slider>
                <Text fontSize='xs' color='grey5'>
                    50%
                </Text>
            </Box>
            <Text fontSize='xs' color='grey5' fontWeight='700'>
                Current interest rate: {state.calculations.interestRate}%
            </Text>
        </Box>
    )
}
