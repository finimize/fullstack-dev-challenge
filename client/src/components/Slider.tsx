import React from 'react'
import {
    Slider as ChakraSlider,
    SliderTrack,
    SliderFilledTrack,
    SliderThumb,
    SliderProps,
} from '@chakra-ui/react'

const Slider = (props: SliderProps) => (
    <ChakraSlider {...props} colorScheme="primary">
        <SliderTrack>
            <SliderFilledTrack bg="primary" />
        </SliderTrack>
        <SliderThumb />
    </ChakraSlider>
)

export default Slider
