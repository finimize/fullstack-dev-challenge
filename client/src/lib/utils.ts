import type { ChartTooltipItem } from 'chart.js'
import { SAVINGS_DEFAULTS } from './constants'

export const createQueryString = (queries: { [key: string]: any }): string =>
    Object.keys(queries)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(queries[key]))
        .join('&')

export const appendYearsQuery = (queryString: string) =>
    `${queryString}&years=${SAVINGS_DEFAULTS.YEARS}`

export const extractYearlyData = (projections: number[]) =>
    projections.filter((item, index) => index % 12 === 0)

export const numberFormatter = new Intl.NumberFormat('gb-GB', {
    style: 'currency',
    currency: 'GBP',
})

export const getTooltipTitle = (tooltipItems: ChartTooltipItem[]) => `Year ${tooltipItems[0].label}`

export const getTooltipLabel = (tooltipItem: ChartTooltipItem) => {
    const value = tooltipItem.value ? parseFloat(tooltipItem.value) : 0
    return numberFormatter.format(value)
}
