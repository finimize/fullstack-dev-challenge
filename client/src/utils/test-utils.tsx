import React, { ReactElement, FC, useState } from 'react'
import { ChakraProvider, extendTheme } from '@chakra-ui/react'
import { render as rtlRender, RenderResult, RenderOptions } from '@testing-library/react'
import { AppContext } from '../store'
import { initialState } from '../store/initialState'
import { StateInterface } from '../store/store.interface'
import { theme } from '../theme'

const mainTheme = extendTheme(theme)

interface DistpatchMockinterface {
    payload: string
    type: string
}

interface CustomRenderOptions extends Partial<RenderOptions> {
    store?: StateInterface
    defaultTheme?: typeof theme
}

interface CustomRenderInterface {
    (
        ui: React.ReactElement<unknown, string | React.JSXElementConstructor<unknown>>,
        options: CustomRenderOptions,
    ): { render: RenderResult; dispatchMock: jest.Mock<unknown, unknown[]> }
}

const customRender: CustomRenderInterface = (
    ui: ReactElement,
    { store = initialState, defaultTheme = mainTheme, ...options },
) => {
    const dispatchMock = jest.fn()
    const AllProviders: FC = ({ children }) => {
        const [calculatorState, setCalculatorState] = useState(store)
        return (
            <ChakraProvider theme={defaultTheme}>
                <AppContext.Provider
                    value={{
                        state: calculatorState,
                        dispatch: dispatchMock.mockImplementation(
                            (value: DistpatchMockinterface) => {
                                setCalculatorState(() => ({
                                    ...calculatorState,
                                    calculatorMode: value.payload,
                                }))
                            },
                        ),
                    }}
                >
                    {' '}
                    {children}{' '}
                </AppContext.Provider>{' '}
            </ChakraProvider>
        )
    }
    return { render: rtlRender(ui, { wrapper: AllProviders, ...options }), dispatchMock }
}

export { customRender }
