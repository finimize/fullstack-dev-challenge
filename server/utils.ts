const compoundedInterest = (interestRate: number, timeInMonths: number) :number => {
    return Math.pow((interestRate + 1), timeInMonths);
};

const percentageToDecimal = (percentValue: number) :number => percentValue/100;

export const monthlyCompoundInterest = (monthlyDeposit: number, monthlyInterest: number, timeInMonths: number) :number =>  {
    const interestRate = percentageToDecimal(monthlyInterest);
    const interestOverMonths = compoundedInterest(interestRate, timeInMonths);
    const monthlyCompound = (monthlyDeposit *((interestOverMonths-1)/interestRate));
    return Math.round(monthlyCompound);
}

export const principleCompoundInterest = (initialDeposit: number,
                                          monthlyInterest: number,
                                          timeInMonths: number) : number => {
    const interestOverMonths = compoundedInterest(percentageToDecimal(monthlyInterest), timeInMonths);
    return Math.round(initialDeposit * interestOverMonths);
}

export const totalSavingsOverTime = (initialDeposit: number,
                                    monthlyDeposit: number,
                                    monthlyInterest: number,
                                    timeInMonths: number) :number => {
    const principle = principleCompoundInterest(initialDeposit, monthlyInterest, timeInMonths);
    const  monthly = monthlyCompoundInterest(monthlyDeposit, monthlyInterest, timeInMonths);
    return monthly+principle;
}

const toRange = (value: number) => {
    let array = [];
    for(let i =1;i<=value;i++){
        array.push(i);
    }
    return array;
}
export const toYearsFromNow = (years: number) :Array<string> => {
    let yearsArray = toRange(years);
    return yearsArray.map(number => {
        let d = new Date();
        let currentYear = d.getFullYear();
        let month = d.getMonth();
        let day = d.getDate();
        let c = new Date(currentYear + number, month, day);
        return c.getFullYear().toString();
    })
}
export const yearlySavingsOverTime = (initialDeposit: number,
                                        monthlyDeposit: number,
                                         monthlyInterest: number,
                                         timeInYears: number) : Array<number> => {
        const arr = toRange(timeInYears);
        const principleCompound = principleCompoundInterest(initialDeposit,monthlyInterest,timeInYears*12);
        return arr.map(number => monthlyCompoundInterest(monthlyDeposit,monthlyInterest,number*12) + principleCompound);
}
