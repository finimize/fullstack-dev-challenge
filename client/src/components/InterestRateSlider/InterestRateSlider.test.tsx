import React from 'react'
import { cleanup, fireEvent } from '@testing-library/react'
import { customRender } from '../../utils/test-utils'
import { InterestRateSlider } from './Slider'

afterEach(cleanup)

describe('component/Slider', () => {
    it('renders Slidercomponent', () => {
        const { render } = customRender(<InterestRateSlider />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })
    it('dispatch on change', () => {
        const { render, dispatchMock } = customRender(<InterestRateSlider />, {})
        const { getByTestId } = render
        const sliderComponent = getByTestId('interest-rate-slider')
        fireEvent.focus(sliderComponent)
        fireEvent.mouseDown(sliderComponent)
        expect(dispatchMock).toHaveBeenCalled()
    })
})
