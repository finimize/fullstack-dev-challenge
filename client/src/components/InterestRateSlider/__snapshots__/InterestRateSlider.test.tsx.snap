// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/Slider renders Slidercomponent 1`] = `
<div>
   
  <div
    class="css-uarar0"
  >
    <label
      class="chakra-form__label css-wuh3rg"
    >
      Interest Rate
    </label>
    <div
      class="css-x3oexv"
    >
      <p
        class="chakra-text css-1njcsuf"
      >
        0%
      </p>
      <div
        class="chakra-slider css-1jzw7c4"
        style="position: relative; user-select: none; outline: 0; padding-top: 0px; padding-bottom: 0px;"
        tabindex="-1"
      >
        <div
          class="chakra-slider__track css-yobz93"
          id="slider-track-1"
          style="position: absolute; top: 50%; transform: translateY(-50%); width: 100%;"
        >
          <div
            class="css-fsxq0m"
          />
          <div
            class="chakra-slider__filled-track css-lbpcfc"
            style="position: absolute; top: 50%; transform: translateY(-50%); width: 40%; left: 0px;"
          />
        </div>
        <div
          aria-orientation="horizontal"
          aria-valuemax="50"
          aria-valuemin="0"
          aria-valuenow="20"
          class="chakra-slider__thumb css-t6tax2"
          data-testid="interest-rate-slider"
          id="slider-thumb-1"
          role="slider"
          style="position: absolute; user-select: none; left: calc(40% - 0px);"
          tabindex="0"
        />
        <input
          type="hidden"
          value="20"
        />
      </div>
      <p
        class="chakra-text css-1njcsuf"
      >
        50%
      </p>
    </div>
    <p
      class="chakra-text css-1as9u30"
    >
      Current interest rate: 
      <b>
        20
        %
      </b>
    </p>
  </div>
   
   
</div>
`;
